### PR TITLE
nixos/ssh: default askpass programs

### LINKED ISSUE #75867: Conflicting definitions for programs.ssh.askPassword in plasma5 and seahorse
**Describe the bug**

When plasma5 and gnome3 are enabled on my configuration.nix I get the following error:
```
# nixos-rebuild boot 2>&1 | tee /var/log/nixos-rebuild.log
building Nix...
building the system configuration...
trace: warning: mlocate does not support searching as user other than root
error: The option `programs.ssh.askPassword' has conflicting definitions, in `/nix/var/nix/profiles/per-user/root/channels/nixos/nixos/modules/services/x11/desktop-managers/plasma5.nix' and `/nix/var/nix/profiles/per-user/root/channels/nixos/nixos/modules/programs/seahorse.nix'.
(use '--show-trace' to show detailed location information)
```

**Metadata**
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.3, NixOS, 20.03pre205624.8636580d6f4 (Markhor)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.1`
 - channels(root): `"nixos-20.03pre205710.352f030b715"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module: services/x11/desktop-managers/plasma5.nix programs/seahorse.nix
```

### PR DESCRIPTION
**Pull request by worldofpeace**
###### Motivation for this change
Set the askpass program for seahorse and plasma5 to prevent #75867

Fixes #75867

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/programs/ssh.nix [hunk 1]
```diff
@@ -8,13 +8,27 @@ let
...
+
+  defaultAskPassword =
+    # Use neutral askpass program if both are enabled
+    if config.programs.seahorse.enable && xserverCfg.desktopManager.plasma5.enable then
+      x11AskPass
+    else if config.programs.seahorse.enable then
+      # Seahorse ssh-askpass is used in GNOME3 and other gtk desktops
+      "${pkgs.gnome3.seahorse}/libexec/seahorse/ssh-askpass"
+    else if xserverCfg.desktopManager.plasma5.enable then
+      "${plasma5.ksshaskpass.out}/bin/ksshaskpass"
```
Comment (jtojnar) [inline nixos/modules/programs/ssh.nix:23]: Missing `pkgs.`?

**Comment by jtojnar** [comment]
Also related https://github.com/NixOS/nixpkgs/issues/24311. Perhaps we should just close it if people can just overwrite the option.

**Comment by worldofpeace** [comment]
> Also related #24311. Perhaps we should just close it if people can just overwrite the option.

Yeah, I would for sure close it because overwriting makes sense as one solution.

**Comment by jtojnar** [comment]
Should not that option be nullable, then?

**Comment by peterhoeg** [comment]
I don't think this is the right way to go about handling this problem.

If I choose one or more proper full-fledged DEs, I would expect at least one of them to take precedence and not a generic askpass.

There are basically 2 "Correct(tm)" ways to do this:

1. don't set it globally, but instead source it as part of the session setup for each DE
2. have a global toggle for preferred DE. There are other applications where each DE provides their own, so being able to globally toggle which DE has the higher priority is an alternative

**Comment by worldofpeace** [comment]
> If I choose one or more proper full-fledged DEs, I would expect at least one of them to take precedence and not a generic askpass.

The question of one taking more precedence is more of a perspective of how strange it's going to look if seahorse's askpass is used in plasma or vice versa. I think in that case it's better to just use the default and let the user choose which one they want to use globally. 

> don't set it globally, but instead source it as part of the session setup for each DE

I don't understand this well. the most elegant solution I thought up of was option two in https://github.com/NixOS/nixpkgs/issues/75867#issuecomment-778693934, which is complicated for a niche condition that just achieves what program to use to ask for someone's ssh credentials. I just don't see any huge payoff to implementing per-session controls to configure this. 

> have a global toggle for preferred DE. There are other applications where each DE provides their own, so being able to globally toggle which DE has the higher priority is an alternative

I really don't understand this either. It sounds like I'd have to introduce an option `preferredDesktop`, keep this code and have it be aware of this. And maybe `preferredDesktop` is useful in other places that need to intuit if someone is using multiple desktops. Major drawback is its uses seem to be broader than this problem itself which would need a release note and probably documentation.

**Pull request closed without merging**